File: jscomp/others/belt_HashMapInt.ml
# 1 "hashmap.cppo.ml"
(***********************************************************************)
(*                                                                     *)
(*                                OCaml                                *)
(*                                                                     *)
(*            Xavier Leroy, projet Cristal, INRIA Rocquencourt         *)
(*                                                                     *)
(*  Copyright 1996 Institut National de Recherche en Informatique et   *)
(*  en Automatique.  All rights reserved.  This file is distributed    *)
(*  under the terms of the GNU Library General Public License, with    *)
(*  the special exception on linking described in file ../LICENSE.     *)
(*                                                                     *)
(***********************************************************************)
(**  Adapted by Authors of BuckleScript 2017                           *)

# 23 "hashmap.cppo.ml"
type key = int
type seed = int
external caml_hash_mix_int : seed -> int -> seed  = "caml_hash_mix_int"
external final_mix : seed -> seed = "caml_hash_final_mix"
let hash (s : key) = 
  final_mix (caml_hash_mix_int 0 s)

# 33 "hashmap.cppo.ml"
module N = Belt_internalBuckets
module C = Belt_internalBucketsType
module A = Belt_Array


type 'b t = (unit, unit, key,'b) N.t


let rec copyBucketReHash  ~h_buckets ~ndata_tail  old_bucket = 
  match C.toOpt old_bucket with 
  | None -> ()
  | Some cell ->
    let nidx = hash (N.key cell) land (A.length h_buckets - 1) in 
    let v = C.return cell in 
    begin match C.toOpt (A.getUnsafe ndata_tail nidx) with
      | None -> 
        A.setUnsafe h_buckets nidx  v
      | Some tail ->
        N.nextSet tail v  (* cell put at the end *)            
    end;          
    A.setUnsafe ndata_tail nidx  v;
    copyBucketReHash  ~h_buckets ~ndata_tail  (N.next cell)


let resize  h =
  let odata = C.buckets h in
  let osize = A.length odata in
  let nsize = osize * 2 in
  if nsize >= osize then begin (* no overflow *)
    let h_buckets = A.makeUninitialized nsize  in
    let ndata_tail = A.makeUninitialized nsize  in (* keep track of tail *)
    C.bucketsSet h  h_buckets;          (* so that indexfun sees the new bucket count *)
    for i = 0 to osize - 1 do
      copyBucketReHash  ~h_buckets ~ndata_tail  (A.getUnsafe odata i)
    done;
    for i = 0 to nsize - 1 do
      match C.toOpt (A.getUnsafe ndata_tail i) with
      | None -> ()
      | Some tail -> N.nextSet tail C.emptyOpt
    done
  end


let rec replaceInBucket  (key : key) info cell = 
    if  N.key cell = key 
    then
      begin
        N.valueSet cell info;
        false
      end
    else
    match C.toOpt (N.next cell) with 
    | None -> true 
    | Some cell -> 
      replaceInBucket key info cell

let set  h (key : key) value =
  let h_buckets = C.buckets h in 
  let buckets_len = A.length h_buckets in 
  let i = hash key land (buckets_len - 1) in 
  let l = A.getUnsafe h_buckets i in  
  (match C.toOpt l with 
  | None -> 
    A.setUnsafe h_buckets i (C.return (N.bucket ~key ~value ~next:C.emptyOpt));
    C.sizeSet h (C.size h + 1);
  | Some bucket -> 
      if replaceInBucket key value bucket then begin
        A.setUnsafe h_buckets i (C.return (N.bucket ~key ~value ~next:l));
        C.sizeSet h (C.size h + 1);
      end   
    );
    if C.size h > buckets_len lsl 1 then resize h 



let rec removeInBucket h h_buckets  i (key : key) prec buckets =
  match C.toOpt buckets with
  | None -> ()
  | Some cell  ->
    let cell_next = N.next cell in 
    if  N.key cell = key 
    then 
      begin
        N.nextSet prec cell_next;
        C.sizeSet h (C.size h - 1);        
      end
    else removeInBucket  h h_buckets i key cell cell_next

let remove  h key =  
  let h_buckets = C.buckets h in 
  let i = hash key land (A.length h_buckets - 1) in  
  let bucket = (A.getUnsafe h_buckets i) in 
  match C.toOpt bucket with 
  | None -> ()
  | Some cell -> 
    if N.key cell = key then 
    begin 
      A.setUnsafe h_buckets i (N.next cell);
      C.sizeSet h (C.size h - 1)
    end 
    else 
      removeInBucket  h h_buckets i key cell (N.next cell)



let rec getAux  (key : key) buckets = 
  match C.toOpt buckets with 
  | None ->
    None
  | Some cell ->
    if key = N.key cell  then Some (N.value cell) 
    else getAux key  (N.next cell)

let get  h (key : key) =
  let h_buckets = C.buckets h in 
  let nid = hash key  land (A.length h_buckets - 1) in 
  match C.toOpt @@ A.getUnsafe h_buckets nid with
  | None -> None
  | Some cell1 ->
    if key = (N.key cell1)  then Some (N.value cell1) else
      match C.toOpt (N.next cell1) with
      | None -> None
      | Some cell2 ->
        if  key = (N.key cell2)  then Some (N.value cell2) else
          match C.toOpt (N.next cell2) with
          | None -> None
          | Some cell3 ->
            if  key = (N.key cell3)  then Some (N.value cell3)
            else getAux  key (N.next cell3)


let rec memInBucket (key : key) cell = 
    (N.key cell)  = key  ||
    (match C.toOpt (N.next cell) with 
    | None -> false
    | Some nextCell -> memInBucket  key nextCell
     )
    
let has h key =
  let h_buckets = C.buckets h in 
  let nid = hash key land (A.length h_buckets - 1) in 
  let bucket = A.getUnsafe h_buckets nid in 
  match C.toOpt bucket with 
  | None -> false
  | Some bucket -> 
    memInBucket  key bucket


let make ~hintSize = C.make ~hintSize ~hash:() ~eq:()
let clear = C.clear
let size = C.size
let forEachU = N.forEachU 
let forEach = N.forEach
let reduceU = N.reduceU
let reduce = N.reduce               
let logStats = N.logStats
let keepMapInPlaceU = N.keepMapInPlaceU 
let keepMapInPlace = N.keepMapInPlace
let toArray = N.toArray
let copy = N.copy
let keysToArray = N.keysToArray
let valuesToArray = N.valuesToArray
let getBucketHistogram = N.getBucketHistogram 
let isEmpty = C.isEmpty

let fromArray arr = 
  let len = A.length arr in 
  let v = make len in 
  for i = 0 to len - 1 do 
    let k,value = (A.getUnsafe arr i) in
    set v k value
  done ;
  v

(* TOOD: optimize heuristics for resizing *)  
let mergeMany h arr =   
  let len = A.length arr in 
  for i = 0 to len - 1 do 
    let k,v = (A.getUnsafe arr i) in
    set h k v 
  done

File: lib/es6/Belt_HashMapInt.js


import * as Primitive_hash from "./Primitive_hash.js";
import * as Primitive_option from "./Primitive_option.js";
import * as Belt_internalBuckets from "./Belt_internalBuckets.js";
import * as Belt_internalBucketsType from "./Belt_internalBucketsType.js";

function copyBucketReHash(h_buckets, ndata_tail, _old_bucket) {
  while (true) {
    let old_bucket = _old_bucket;
    if (old_bucket === undefined) {
      return;
    }
    let nidx = Primitive_hash.hash_final_mix(Primitive_hash.hash_mix_int(0, old_bucket.key)) & (h_buckets.length - 1 | 0);
    let tail = ndata_tail[nidx];
    if (tail !== undefined) {
      tail.next = old_bucket;
    } else {
      h_buckets[nidx] = old_bucket;
    }
    ndata_tail[nidx] = old_bucket;
    _old_bucket = old_bucket.next;
    continue;
  };
}

function replaceInBucket(key, info, _cell) {
  while (true) {
    let cell = _cell;
    if (cell.key === key) {
      cell.value = info;
      return false;
    }
    let cell$1 = cell.next;
    if (cell$1 === undefined) {
      return true;
    }
    _cell = cell$1;
    continue;
  };
}

function set(h, key, value) {
  let h_buckets = h.buckets;
  let buckets_len = h_buckets.length;
  let i = Primitive_hash.hash_final_mix(Primitive_hash.hash_mix_int(0, key)) & (buckets_len - 1 | 0);
  let l = h_buckets[i];
  if (l !== undefined) {
    if (replaceInBucket(key, value, l)) {
      h_buckets[i] = {
        key: key,
        value: value,
        next: l
      };
      h.size = h.size + 1 | 0;
    }
    
  } else {
    h_buckets[i] = {
      key: key,
      value: value,
      next: undefined
    };
    h.size = h.size + 1 | 0;
  }
  if (h.size > (buckets_len << 1)) {
    let odata = h.buckets;
    let osize = odata.length;
    let nsize = (osize << 1);
    if (nsize < osize) {
      return;
    }
    let h_buckets$1 = new Array(nsize);
    let ndata_tail = new Array(nsize);
    h.buckets = h_buckets$1;
    for (let i$1 = 0; i$1 < osize; ++i$1) {
      copyBucketReHash(h_buckets$1, ndata_tail, odata[i$1]);
    }
    for (let i$2 = 0; i$2 < nsize; ++i$2) {
      let tail = ndata_tail[i$2];
      if (tail !== undefined) {
        tail.next = undefined;
      }
      
    }
    return;
  }
  
}

function remove(h, key) {
  let h_buckets = h.buckets;
  let i = Primitive_hash.hash_final_mix(Primitive_hash.hash_mix_int(0, key)) & (h_buckets.length - 1 | 0);
  let bucket = h_buckets[i];
  if (bucket !== undefined) {
    if (bucket.key === key) {
      h_buckets[i] = bucket.next;
      h.size = h.size - 1 | 0;
      return;
    } else {
      let _prec = bucket;
      let _buckets = bucket.next;
      while (true) {
        let buckets = _buckets;
        let prec = _prec;
        if (buckets === undefined) {
          return;
        }
        let cell_next = buckets.next;
        if (buckets.key === key) {
          prec.next = cell_next;
          h.size = h.size - 1 | 0;
          return;
        }
        _buckets = cell_next;
        _prec = buckets;
        continue;
      };
    }
  }
  
}

function get(h, key) {
  let h_buckets = h.buckets;
  let nid = Primitive_hash.hash_final_mix(Primitive_hash.hash_mix_int(0, key)) & (h_buckets.length - 1 | 0);
  let cell1 = h_buckets[nid];
  if (cell1 === undefined) {
    return;
  }
  if (key === cell1.key) {
    return Primitive_option.some(cell1.value);
  }
  let cell2 = cell1.next;
  if (cell2 === undefined) {
    return;
  }
  if (key === cell2.key) {
    return Primitive_option.some(cell2.value);
  }
  let cell3 = cell2.next;
  if (cell3 !== undefined) {
    if (key === cell3.key) {
      return Primitive_option.some(cell3.value);
    } else {
      let _buckets = cell3.next;
      while (true) {
        let buckets = _buckets;
        if (buckets === undefined) {
          return;
        }
        if (key === buckets.key) {
          return Primitive_option.some(buckets.value);
        }
        _buckets = buckets.next;
        continue;
      };
    }
  }
  
}

function has(h, key) {
  let h_buckets = h.buckets;
  let nid = Primitive_hash.hash_final_mix(Primitive_hash.hash_mix_int(0, key)) & (h_buckets.length - 1 | 0);
  let bucket = h_buckets[nid];
  if (bucket !== undefined) {
    let _cell = bucket;
    while (true) {
      let cell = _cell;
      if (cell.key === key) {
        return true;
      }
      let nextCell = cell.next;
      if (nextCell === undefined) {
        return false;
      }
      _cell = nextCell;
      continue;
    };
  } else {
    return false;
  }
}

function make(hintSize) {
  return Belt_internalBucketsType.make(undefined, undefined, hintSize);
}

function size(h) {
  return h.size;
}

function fromArray(arr) {
  let len = arr.length;
  let v = Belt_internalBucketsType.make(undefined, undefined, len);
  for (let i = 0; i < len; ++i) {
    let match = arr[i];
    set(v, match[0], match[1]);
  }
  return v;
}

function mergeMany(h, arr) {
  let len = arr.length;
  for (let i = 0; i < len; ++i) {
    let match = arr[i];
    set(h, match[0], match[1]);
  }
}

let clear = Belt_internalBucketsType.clear;

let isEmpty = Belt_internalBucketsType.isEmpty;

let copy = Belt_internalBuckets.copy;

let forEachU = Belt_internalBuckets.forEach;

let forEach = Belt_internalBuckets.forEach;

let reduceU = Belt_internalBuckets.reduce;

let reduce = Belt_internalBuckets.reduce;

let keepMapInPlaceU = Belt_internalBuckets.keepMapInPlace;

let keepMapInPlace = Belt_internalBuckets.keepMapInPlace;

let toArray = Belt_internalBuckets.toArray;

let keysToArray = Belt_internalBuckets.keysToArray;

let valuesToArray = Belt_internalBuckets.valuesToArray;

let getBucketHistogram = Belt_internalBuckets.getBucketHistogram;

let logStats = Belt_internalBuckets.logStats;

export {
  make,
  clear,
  isEmpty,
  set,
  copy,
  get,
  has,
  remove,
  forEachU,
  forEach,
  reduceU,
  reduce,
  keepMapInPlaceU,
  keepMapInPlace,
  size,
  toArray,
  keysToArray,
  valuesToArray,
  fromArray,
  mergeMany,
  getBucketHistogram,
  logStats,
}
/* No side effect */
